library/std/src/sys/unix/ext/net/ancillary.rs

```diff
@@ -0,0 +1,666 @@
+            msg_flags: 0,
+        };
+
+        let count = socket.recv_msg(&mut msg)?;
+
+        ancillary.length = msg.msg_controllen;
+        ancillary.truncated = msg.msg_flags & libc::MSG_CTRUNC == libc::MSG_CTRUNC;
+
+        let truncated = msg.msg_flags & libc::MSG_TRUNC == libc::MSG_TRUNC;
+        let addr = SocketAddr::from_parts(msg_name, msg.msg_namelen);
+
+        Ok((count, truncated, addr))
+    }
+}
+
+pub(super) fn send_vectored_with_ancillary_to(
+    socket: &Socket,
+    path: Option<&Path>,
+    bufs: &mut [IoSliceMut<'_>],
+    ancillary: &mut SocketAncillary<'_>,
+) -> io::Result<usize> {
+    unsafe {
+        let (mut msg_name, msg_namelen) =
+            if let Some(path) = path { sockaddr_un(path)? } else { (zeroed(), 0) };
+
+        let mut msg = libc::msghdr {
+            msg_name: &mut msg_name as *mut _ as *mut _,
+            msg_namelen,
+            msg_iov: bufs.as_mut_ptr().cast(),
+            msg_iovlen: bufs.len(),
+            msg_control: ancillary.buffer.as_mut_ptr().cast(),
+            msg_controllen: ancillary.length,
+            msg_flags: 0,
+        };
+
+        ancillary.truncated = false;
+
+        socket.send_msg(&mut msg)
+    }
+}
+
+fn add_to_ancillary_data<T>(
+    buffer: &mut [u8],
+    length: &mut usize,
+    source: &[T],
+    cmsg_level: libc::c_int,
+    cmsg_type: libc::c_int,
+) -> bool {
+    let source_len = if let Some(source_len) = source.len().checked_mul(size_of::<T>()) {
+        if let Ok(source_len) = u32::try_from(source_len) {
+            source_len
+        } else {
+            return false;
+        }
+    } else {
+        return false;
+    };
+
+    unsafe {
+        let additional_space = libc::CMSG_SPACE(source_len) as usize;
+
+        let new_length = if let Some(new_length) = additional_space.checked_add(*length) {
+            new_length
+        } else {
+            return false;
+        };
+
+        if new_length > buffer.len() {
+            return false;
+        }
+
+        buffer[*length..new_length].fill(0);
+
+        *length = new_length;
+
+        let msg = libc::msghdr {
+            msg_name: null_mut(),
+            msg_namelen: 0,
+            msg_iov: null_mut(),
+            msg_iovlen: 0,
+            msg_control: buffer.as_mut_ptr().cast(),
+            msg_controllen: *length,
+            msg_flags: 0,
+        };
+
+        let mut cmsg = libc::CMSG_FIRSTHDR(&msg);
+        let mut previous_cmsg = cmsg;
+        while !cmsg.is_null() {
+            previous_cmsg = cmsg;
+            cmsg = libc::CMSG_NXTHDR(&msg, cmsg);
+        }
+
+        if previous_cmsg.is_null() {
+            return false;
+        }
+
+        (*previous_cmsg).cmsg_level = cmsg_level;
+        (*previous_cmsg).cmsg_type = cmsg_type;
+        (*previous_cmsg).cmsg_len = libc::CMSG_LEN(source_len) as usize;
+
+        let data = libc::CMSG_DATA(previous_cmsg).cast();
+
+        libc::memcpy(data, source.as_ptr().cast(), source_len as usize);
```

Comment:
This is still using a libc::memcpy, instead of either ptr::copy_nonoverlapping or ptr::copy.

We also need a `T: Copy` bound here, I suspect, since use cases with non-Copy data are likely to be wrong.

Reply:
Thanks for the suggestion. Unfortunately, I cannot use `ptr::copy_nonoverlapping` or `ptr::copy`, because these functions required that the destination address is properly aligned. `man 3 cmsg`:
```
       *  CMSG_DATA()  returns  a  pointer  to  the  data  portion of a cmsghdr.  The
          pointer returned cannot be assumed to be suitably aligned for accessing ar‐
          bitrary  payload  data types.  Applications should not cast it to a pointer
          type matching the payload, but should instead use memcpy(3) to copy data to
          or from a suitably declared object.
```

Reply:
You’re already casting the pointer away to `*mut u8` (which is always aligned) for `memcpy`. You can do exactly the same for `ptr::copy{,_nonoverlapping}`.

Reply:
If I use `u8` as `T` then what are the benefits between `libc::memcpy` and `ptr::copy`?